components/script/dom/bindings/structuredclone.rs

```diff
@@ -5,15 +5,148 @@
 use dom::bindings::error::{Error, Fallible};
+use dom::bindings::js::Root;
+use dom::bindings::reflector::DomObject;
+use dom::blob::{Blob, BlobImpl};
 use dom::globalscope::GlobalScope;
-use js::jsapi::{HandleValue, MutableHandleValue};
-use js::jsapi::{JSContext, JS_ReadStructuredClone, JS_STRUCTURED_CLONE_VERSION};
-use js::jsapi::{JS_ClearPendingException, JS_WriteStructuredClone};
+use js::jsapi::{Handle, HandleObject, HandleValue, MutableHandleValue, JSAutoCompartment, JSContext};
+use js::jsapi::{JSStructuredCloneCallbacks, JSStructuredCloneReader, JSStructuredCloneWriter};
+use js::jsapi::{JS_ClearPendingException, JSObject, JS_ReadStructuredClone};
+use js::jsapi::{JS_ReadBytes, JS_WriteBytes};
+use js::jsapi::{JS_ReadUint32Pair, JS_WriteUint32Pair};
+use js::jsapi::{JS_STRUCTURED_CLONE_VERSION, JS_WriteStructuredClone};
+use js::jsapi::{MutableHandleObject, TransferableOwnership};
 use libc::size_t;
+use std::os::raw;
 use std::ptr;
 use std::slice;
 
+// TODO: Should we add Min and Max const to https://github.com/servo/rust-mozjs/blob/master/src/consts.rs?
+// TODO: Determine for sure which value Min and Max should have.
+// NOTE: Current values found at https://dxr.mozilla.org/mozilla-central/
+// rev/ff04d410e74b69acfab17ef7e73e7397602d5a68/js/public/StructuredClone.h#323
+enum StructuredCloneTags {
+    /// To support additional types, add new tags with values incremented from the last one before Max.
+    Min = 0xFFFF8000,
+    DomBlob = 0xFFFF8001,
+    Max = 0xFFFFFFFF,
+}
+
+unsafe fn write_length(w: *mut JSStructuredCloneWriter,
+                       length: usize) {
+  let high: u32 = (length >> 32) as u32;
+  let low: u32 = length as u32;
+  assert!(JS_WriteUint32Pair(w, high, low));
+}
+
+unsafe fn read_length(r: *mut JSStructuredCloneReader)
+                      -> usize {
+  let mut high: u32 = 0;
+  let mut low: u32 = 0;
+  assert!(JS_ReadUint32Pair(r, &mut high as *mut u32, &mut low as *mut u32));
+  return (low << high) as usize;
+}
+
+unsafe fn read_blob(cx: *mut JSContext,
+                    r: *mut JSStructuredCloneReader)
+                    -> *mut JSObject {
+    let blob_length = read_length(r);
+    let type_str_length = read_length(r);
+    let mut blob_buffer = vec![0u8; blob_length];
+    assert!(JS_ReadBytes(r, blob_buffer.as_mut_ptr() as *mut raw::c_void, blob_length));
+    let mut type_str_buffer = vec![0u8; type_str_length];
+    assert!(JS_ReadBytes(r, type_str_buffer.as_mut_ptr() as *mut raw::c_void, type_str_length));
+    let type_str = String::from_raw_parts(type_str_buffer.as_ptr() as *mut _, type_str_length, type_str_length);
+    let target_global = GlobalScope::from_context(cx);
+    let blob = Blob::new(&target_global, BlobImpl::new_from_bytes(blob_buffer), type_str);
+    return blob.reflector().get_jsobject().get()
+}
+
+unsafe fn write_blob(blob: Root<Blob>,
+                     w: *mut JSStructuredCloneWriter)
+                     -> bool {
```

Comment:
thanks, I was about to ask you if the code needed to do anything in case of an error, and then read about the early return of `try!`, interesting macro that is...